SWI-cpp2.h

```diff
@@ -1515,4 +1516,144 @@ template <typename ContextType> class PlForeignContextPtr
+		 *          BLOBS		*
+		 *******************************/
+
+// TODO: these should be in a namespace
+
+
+template<typename C_t> [[nodiscard]]
+static C_t *
+cast_blob(PlAtom aref)
+{ size_t len;
+  PL_blob_t *type;
+  auto ref = static_cast<C_t *>(aref.blob_data(&len, &type));
+  if ( ref && type == ref->blob_t_)
+  { assert(len == sizeof *ref);
+    assert(ref->validate());
+    return ref;
+  }
+  return nullptr;
+}
+
+template<typename C_t> [[nodiscard]]
+static C_t*
+cast_blob_check(PlAtom aref)
+{ auto ref = cast_blob<C_t>(aref);
+  assert(ref);
+  return ref;
+}
+
+template<typename C_t>
+void blob_acquire(atom_t a)
+{ PlAtom a_(a);
+  auto data = cast_blob_check<C_t>(a_);
+  data->acquire(a_);
+}
+
+template<typename C_t> [[nodiscard]]
+int blob_release(atom_t a)
+{ auto data = cast_blob_check<C_t>(PlAtom(a));
+  delete data;
+  return true;
+}
+
+template<typename C_t> [[nodiscard]]
+int blob_compare(atom_t a, atom_t b)
+{ const auto a_data = cast_blob_check<C_t>(PlAtom(a));
+  const auto b_data = cast_blob_check<C_t>(PlAtom(b));
+  int rc = a_data->compare_fields(b_data);
+  if ( rc == 0 )
+    { return (a_data < b_data) ? -1 : (a_data > b_data) ? 1 : 0;
+  }
+  return rc;
+}
+
+template<typename C_t> [[nodiscard]]
+int blob_write(IOSTREAM *s, atom_t a, int flags)
+{ const auto data = cast_blob_check<C_t>(PlAtom(a));
+  return data->write(s, flags);
+}
+
+template<typename C_t> [[nodiscard]]
+int blob_save(atom_t a, IOSTREAM *fd)
+{ const auto data = cast_blob_check<C_t>(PlAtom(a));
+  return data->save(fd);
+}
+
+template<typename C_t> [[nodiscard]]
+atom_t blob_load(IOSTREAM *fd)
+{ return C_t::load(fd).C_;
+}
+
+template<const PL_blob_t& blob_t>
+class PlBlob
+{
+public:
+  explicit PlBlob()
+    : blob_t_(&blob_t),
+      symbol_(PlAtom(PlAtom::null)) // filled in by acquire()
+  { }
+  explicit PlBlob(const PlBlob&) = delete;
+  explicit PlBlob(PlBlob&&) = delete;
+  PlBlob& operator =(const PlBlob&) = delete;
+  virtual ~PlBlob() = default;
+
+  bool validate() const { return blob_t_ == &blob_t; }
+
+  void acquire(PlAtom _symbol) { symbol_ = _symbol; }
+
+  bool symbol_not_null() const { return symbol_.not_null(); }
+
+  PlTerm symbol_term() const;
+
+  virtual int compare_fields(const PlBlob<blob_t> *_b) const
+  { return 0;
+  }
+
+  bool write(IOSTREAM *s, int flags) const;
+
+  bool virtual write_fields(IOSTREAM *s, int flags) const
+  { return true; }
+
+  bool save(IOSTREAM *fd) const;
+
+  static PlAtom load(IOSTREAM *fd);
+
+  const PL_blob_t *blob_t_;
+  PlAtom symbol_;		/* associated symbol (used for error terms) */
+};
+
+
+template<const PL_blob_t& blob_t>
+inline bool PlBlob<blob_t>::write(IOSTREAM *s, int flags) const
+{ return Sfprintf(s, "<%s>(%p", blob_t.name, this) &&
+    write_fields(s, flags) &&
+    // TODO: following should write ")\n" if PL_WRT_NEWLINE but there
+    //       appears to be a bug in the top-level code that sets the
+    //       flag:
+    Sfprintf(s, PL_WRT_NEWLINE ? ")" : ")");
```

Comment:
PL_WRT_NEWLINE has very terse documentation. Does it mean that the write callback is being called in the context of a top-level display of results? - because it doesn't seem to mean "put a newline after this", both by experimentation and by reading the code.

If so, I'll change this code (and also similar code in library(pcre) and possibly elsewhere).

Reply:
The flag is passed, but doesn't expect any action from the blob handler.    It does cause PL_write_term() to add a newline at the end.   This is convenient as it allows output with a newline in a single call.   Using a single call, no other output (from threads) can be between, so it also avoids having to lock the stream explicitly.

Most of the flags are meaningless.   The writer hook may decide to act on quoted write.

Reply:
So, if I call PL_write_term() from inside the blob's "write" callback, I should use `flags & (~PL_WRT_NEWLINE)`? That would propagate, e.g. `PL_WRT_QUOTED`.

I've changed the code from
```cpp
    if ( pathname.not_null() )
      if ( !Sfprintf(s, ",path=%Ws", pathname.as_wstring().c_str()) )
         return false;
```
to
```cpp
    flags &= ~PL_WRT_NEWLINE;
    if ( pathname.not_null() )
      if ( !Sfprintf(s, ",path=") ||
           !PlTerm_atom(pathname).write(s, 1200, flags) ) // calls PL_term_write()
         return false;
``

Reply:
I would typically not call Prolog write anyway.   Surely I've never needed it.   In this case, I'd get the text from the atom as UTF-8 and use `%Us` with Sfprintf().  If you really want, I guess you can.   Only the `release` hook does not allow for callbacks as it will get atom-gc into a deadlock.    If you do a write callback you probably do not want to pass on any flags, possibly except a few specific ones.